### ISSUE TITLE
Wagtail v 4.2.4 Conflicting migrations

### ISSUE THREAD

**Issue by nickmoreton** [wagtail/wagtail#10724]
### Issue Summary

Testing fails when testing with the specific Wagtail 4.2.4 version

I have raised this issue here: https://github.com/cfpb/wagtail-sharing/issues/66 as it was the package I was working on when I discovered the issue.

I'm not sure it's an issue with the package but more of an issue with v4.2.4

I'm looking for some help to decide.

### Steps to Reproduce

I have described the steps to reproduce it here: https://github.com/cfpb/wagtail-sharing/issues/66

### Technical details

-   Python version: 3.8 to 3.11
-   Django version: 3.2 to 4.1
-   Wagtail version: 4.2 (4.2.4 specifically)
-   Browser version: N/A

**Comment by chosak** [comment]
@nickmoreton I agree that it looks like this is an issue with the Wagtail 4.2.4 release, perhaps the way the PyPI build artifact was created. If you `pip install wagtail==4.2.3`, the testapp migrations folder includes migrations 0001-0019, as is consistent with the [tagged repository state for 4.2.3](https://github.com/wagtail/wagtail/tree/v4.2.3/wagtail/test/testapp/migrations).

If you `pip install wagtail==4.2.4`, though, the testapp migrations folder includes both the 0001-0019 migrations from [the tagged repository state for 4.2.4](https://github.com/wagtail/wagtail/tree/v4.2.4/wagtail/test/testapp/migrations) but also the unsquashed migrations 0001-0073 from [the tagged repository state for 4.1.6](https://github.com/wagtail/wagtail/tree/v4.1.6/wagtail/test/testapp/migrations). So the 4.2.4 release is including files it shouldn't be.

The duplicate migration files is what causes issues for any Wagtail project that tries to include testapp; this maybe isn't something that is officially supported or recommended but seems like a reasonable practice, for example, wagtail-sharing does this to make use of the convenient test models in that app.

I imagine this wasn't picked up in CI because CI always tests against the repository state where this issue seems like it had to do with the local state of the machine that created the release artifact for PyPI. [This wiki page](https://github.com/wagtail/wagtail/wiki/Creating-a-new-Wagtail-release#package-build) suggests that it is a manual process and recommends:

> If you're creating multiple releases (e.g. patch releases for different minor releases), create a fresh clone for each release. This ensures that we do not keep any files that have been moved/deleted.

Is it possible that maybe this didn't happen with 4.2.4 -- maybe 4.1.6 was built and then 4.2.4 was built in such a way that those squashed migration files accidentally got included in the package?

**Comment by Stormheg** [comment]
In my opinion, one should not rely on anything inside of testapp. It is not documented and intended to be used in Wagtails internal test suite only. We can't provide support or guarantee compatibility.

With that said, including those extra migrations looks like a packaging mistake indeed.

We include Wagtails tests and the testapp in releases of Wagtail with the intend that redistributors (if any, I don't know anyone doing this...) can run Wagtails test suite. This was discussed here: https://github.com/wagtail/wagtail/issues/6130#issuecomment-654283698. In that sense, the unsquashed migrations are a problem because they prevent running the testsuite. Still not something I consider warrants fixing.

**Comment by laymonage** [comment]
> > If you're creating multiple releases (e.g. patch releases for different minor releases), create a fresh clone for each release. This ensures that we do not keep any files that have been moved/deleted.
> 
> Is it possible that maybe this didn't happen with 4.2.4 -- maybe 4.1.6 was built and then 4.2.4 was built in such a way that those squashed migration files accidentally got included in the package?

@chosak It's likely that it was the case. 4.1.6, 4.2.4, and 5.0.1 was released at the same time, and I made the mistake of reusing the same clone for those patch releases. I thought git would've cleaned up the removed/renamed files when I switched between the branches. That note in the release guide was added retrospectively after #10522 was reported.

**Comment by chosak** [comment]
@Stormheg @laymonage thanks for the feedback. For now I'm moving to remove our use of testapp -- a quick search of GitHub and I didn't find any other cases of people doing this and I understand that it's not really supported code. I will say that for developing third-party Wagtail packages it can be convenient to use the testapp models instead of rolling your own, especially for tests, although the Django 4.1+ [`isolate_apps`](https://docs.djangoproject.com/en/4.1/topics/testing/tools/#isolating-apps) functionality makes doing that a little bit easier.

**Comment by laymonage** [comment]
Given that 4.2 is well out of support now, it's safe to say we're not going to release a new patch version that removes the problematic migrations. This isn't a blocker for normal setups, and we don't advertise the testapp as being something that packages can rely on, so I think we can close this. Sorry for the trouble, and thank you for reporting!
